clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -0,0 +1,196 @@
+        return new DefaultCacheResult(Collections.emptyMap());
+      }
+
+      CacheData metadata = providerCache.get(getAgentType(), "metadata", RelationshipCacheFilter.none());
+      MetadataAttributes metadataAttributes = null;
+      if (metadata != null) {
+        metadataAttributes = objectMapper.convertValue(metadata.getAttributes(), MetadataAttributes.class);
+      }
+
+      Set<String> instanceTypes = null;
+      if (metadataAttributes != null && metadataAttributes.etag != null && metadataAttributes.cachedInstanceTypes != null) {
+
+        //we have enough from a previous request to not re-request if the etag is unchanged..
+        HttpResponse headResponse = httpClient.execute(pricingHost, new HttpHead(pricingUri));
+        EntityUtils.consumeQuietly(headResponse.getEntity());
+        if (headResponse.getStatusLine().getStatusCode() != 200) {
+          throw new Exception("failed to read instance type metadata for " + region + ": " + headResponse.getStatusLine().toString());
+        }
+
+        Optional<String> etag = getEtagHeader(headResponse);
+
+        if (etag.filter(metadataAttributes.etag::equals).isPresent()) {
+          instanceTypes = metadataAttributes.cachedInstanceTypes;
+        }
+      }
+      if (instanceTypes == null) {
+        HttpResponse getResponse = httpClient.execute(pricingHost, new HttpGet(pricingUri));
+        if (getResponse.getStatusLine().getStatusCode() != 200) {
+          EntityUtils.consumeQuietly(getResponse.getEntity());
+          throw new Exception("failed to read instance type data for " + region + ": " + getResponse.getStatusLine().toString());
+        }
+        Optional<String> etag = getEtagHeader(getResponse);
+        HttpEntity entity = getResponse.getEntity();
+        instanceTypes = fromStream(entity.getContent());
+        EntityUtils.consumeQuietly(entity);
+        if (etag.isPresent()) {
+          metadataAttributes = new MetadataAttributes();
+          metadataAttributes.etag = etag.get();
+          metadataAttributes.cachedInstanceTypes = new HashSet<>(instanceTypes);
+          metadata = new DefaultCacheData("metadata", objectMapper.convertValue(metadataAttributes, ATTRIBUTES), Collections.emptyMap());
+        } else {
+          metadata = null;
+        }
+      }
+      Map<String, Collection<String>> evictions = new HashMap<>();
+      Map<String, Collection<CacheData>> cacheResults = new HashMap<>();
+      List<CacheData> instanceTypeData = new LinkedList<>();
+      cacheResults.put(Keys.Namespace.INSTANCE_TYPES.getNs(), instanceTypeData);
+      if (metadata != null) {
+        cacheResults.put(getAgentType(), Collections.singleton(metadata));
+      } else {
+        evictions.put(getAgentType(), Collections.singleton("metadata"));
+      }
+
+      for (String instanceType : instanceTypes) {
+        for (String account : matchingAccounts) {
+          Map<String, Object> instanceTypeAttributes = new HashMap<>();
```

Comment:
If we have Guava around, I would do 

```
  Map<String, Object> instanceTypeAttributes = ImmutableMap.of(
    "account", account,
    "region", region,
    "name", instanceType);
```

Reply:
I'm always leery of the guava and it's fragility so I tend to avoid it unless I'm going to use a whole lot of it